datafusion/physical-plan/src/topk/mod.rs

```diff
@@ -186,6 +235,90 @@ impl TopK {
         Ok(())
     }
 
+    fn calculate_dynamic_filters(
+        thresholds: Vec<ColumnThreshold>,
+    ) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+        // Create filter expressions for each threshold
+        let mut filters: Vec<Arc<dyn PhysicalExpr>> =
```

Comment:
I wonder: is there any way we could make only the right hand side of `col > 123` "dynamic"? That is: could we return `col > <dynamic thing>` instead of `<dynamic thing>`? I guess we'd need to some sort of `PhysicalExpr` that `BinaryExpr` and others can work with? I don't know how `BinaryExpr` is implemented internally.

Reply:
`col > 123` is implemented like `BinaryExpr(Column, Literal)`

I really think the idea of `DynamicLiteral` or something is quite compelling 

You could store 
```rust
struct DynamicLiteral {
  literal: Arc<Mutex<Arc<ScalarValur>>>>
}
```

And follow the model of evaluation for [`Literal`](https://github.com/apache/datafusion/blob/main/datafusion/physical-expr/src/expressions/literal.rs)

Except you would first clone the Arc before execution

That would mean the `Mutex` is only held long enough to Clone Arcs so would minimize contention

Reply:
What would we initialize it to when we've got 0 items in the TopK heap?

Reply:
Been thinking about this too :D.

One could add another term to the generated expression:
```
passive OR <cur output of calculate_dynamic_filters>
```
`passive` itself is a dynamic literal that starts out as `true` and gets set to `false` by the TopK Heap once the heap is full.

A downside of the DynamicLiteral approach is that it requires one mutex per field (vs 1 mutex for the whole expression previously). For native types, this *feels* like something that could be backed by an atomic, but I don't see an easy way of achieving that.